File: hash/leetcode_36/solution.py
import math

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        col_set = [set() for _ in range(9)]
        row_set = [set() for _ in range(9)]
        box_set = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                cell = board[i][j]
                box_index = (i // 3) * 3 + (j // 3)

                if cell == '.':
                    continue

                if cell in row_set[i]:
                    return False
                else:
                    row_set[i].add(cell)

                if cell in col_set[j]:
                    return False
                else:
                    col_set[j].add(cell)

                if cell in box_set[box_index]:
                    return False
                else:
                    box_set[box_index].add(cell)

        return True

File: hash/leetcode_36/solution.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    const rowSet = Array.from({length: 9}, () => new Set())
    const colSet = Array.from({length: 9}, () => new Set())
    const boxSet = Array.from({length: 9}, () => new Set())

    for (let i=0; i<9; i++) {
        for (let j=0; j<9; j++) {
            const cell = board[i][j];

            if (cell === '.') {
                continue;
            }

            const boxIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3);

            if(rowSet[i].has(cell)) {
                return false;
            } else {
                rowSet[i].add(cell)
            }

            if(colSet[j].has(cell)) {
                return false;
            } else {
                colSet[j].add(cell);
            }

            if(boxSet[boxIndex].has(cell)) {
                return false;
            } else {
                boxSet[boxIndex].add(cell);
            }
        }
    }

    return true;
};